src/Symfony/Component/Translation/Extractor/PhpAstExtractor.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Extractor;
+
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitor;
+use PhpParser\Parser;
+use PhpParser\ParserFactory;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Translation\Extractor\Visitor\ConstraintVisitor;
+use Symfony\Component\Translation\Extractor\Visitor\TranslatableMessageVisitor;
+use Symfony\Component\Translation\Extractor\Visitor\TransMethodVisitor;
+use Symfony\Component\Translation\Extractor\Visitor\Visitor;
+use Symfony\Component\Translation\MessageCatalogue;
+
+/**
+ * PhpAstExtractor extracts translation messages from a PHP AST.
+ *
+ * @author Mathieu Santostefano <msantostefano@protonmail.com>
+ */
+class PhpAstExtractor extends AbstractFileExtractor implements ExtractorInterface
+{
+    private string $prefix = '';
+    private Parser $parser;
+    /**
+     * @var array<Visitor&NodeVisitor>
+     */
+    private array $visitors;
+
+    public function __construct()
+    {
+        $this->parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
+        $this->visitors = [
```

Comment:
no DI for those two props: on purpose as this shouldn't be extensible?

Reply:
It's part of question I have about this feature. IMHO, it's better to not allow extension for the parser (I see no reason for a user to create a parser for PHP5 code, as 6.2 is PHP8.1+ compatible).

For the visitors, it could happen for a user to need a special visitor that we don't have in the core. So, yes I can inject a service locator here.

Reply:
I would keep them as an implementation detail for now.

Reply:
Service locator for visitors could be nice.

If we use another `TranslatableInterface` than `TranslatableMessage` we may want implements cutsom Visitor

Reply:
It's done